File: 0720-longest-word-in-dictionary/0720-longest-word-in-dictionary.ts
class TrieNode {
    children: { [key: string]: TrieNode };
    isEndOfWord: boolean;

    constructor() {
        this.children = {};   // plain object instead of Map
        this.isEndOfWord = false;
    }
}

class Trie {
    root: TrieNode;

    constructor() {
        this.root = new TrieNode();
    }

    insert(word: string): void {
        let current = this.root;
        for (const c of word) {
            if (!(c in current.children)) {
                current.children[c] = new TrieNode();
            }
            current = current.children[c];
        }
        current.isEndOfWord = true;
    }

    canBuild(word: string): boolean {
        let current = this.root;
        for (const c of word) {
            if (!(c in current.children)) {
                return false;
            }
            current = current.children[c];
            if (!current.isEndOfWord) {
                return false;
            }
        }
        return true;
    }
}

function longestWord(words: string[]): string {
    const trie = new Trie();
    for (const word of words) {
        trie.insert(word);
    }

    let answer = "";
    for (const word of words) {
        if (trie.canBuild(word)) {
            if (
                word.length > answer.length ||
                (word.length === answer.length && word < answer)
            ) {
                answer = word;
            }
        }
    }
    return answer;
}

File: 0720-longest-word-in-dictionary/0720-longest-word-in-dictionary.py
from typing import List

class TrieNode:
    def __init__(self):
        self.children = {}  # dictionary: char -> TrieNode
        self.is_end_of_word = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        current = self.root
        for c in word:
            if c not in current.children:
                current.children[c] = TrieNode()
            current = current.children[c]
        current.is_end_of_word = True

    def can_build(self, word: str) -> bool:
        """
        Check if this word can be built one character at a time
        with all prefixes present in the Trie.
        """
        current = self.root
        for c in word:
            if c not in current.children:
                return False
            current = current.children[c]
            if not current.is_end_of_word:
                return False
        return True


class Solution:
    def longestWord(self, words: List[str]) -> str:
        trie = Trie()
        for word in words:
            trie.insert(word)

        answer = ""
        for word in words:
            if trie.can_build(word):
                if len(word) > len(answer) or (len(word) == len(answer) and word < answer):
                    answer = word
        return answer
